californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java

```diff
@@ -193,6 +193,10 @@ public void runStriped() {
 				}
 			}
 
+			@Override
+			public void onSendError(Throwable error) {
+				exchange.executeComplete();
+			}
```

Comment:
I would prefer to move it into the `ExchangeCleanupLayer`. (I changed/edited my comment therefore in the issue):

Reply:
Cleanup on the "ReliabilityLayer" will only work for CON, but NON request may still leak. 
Even if leshan usually doesn't use such NON requests, I would prefer to add the cleanup to the "ExchangeCleanupLayer" extending/add that with a cleanup for "CON responses". 
And I feel, that checking to use the "failed()" callback (instead of "onSendError") )is also a good idea.

Reply:
The real question : "Where to add the exchange.executeComplete?"

About NON, CON and ReliabilityLayer, I agree.

About ExchanCleanupLayer, I suspect there is a potential problem because request send by blockwiseLayer will no go through this layer. :(

A good practice :  The class which add the element in the store should be responsible to remove it.
So I would add a CleanerMessageObserver before to add in the store. This way we be sure that all element added has a cleaner attached. WDYT ?

Reply:
AFAIK, the `MessageObserver` are copied from the outer requests to the single blocks requests. So I think, it should work with `ExchangeCleanupLayer`.
And Exchanges for notifies are not stored in the `InMemoryMessageExchangeStore`. 
But in the end, I guess without testing the blockwise & notifies, we will not be sure, if the leak is fixed, regardless which implementation is chosen.

Reply:
> But in the end, I guess without testing the blockwise & notifies, we will not be sure, if the leak is fixed, regardless which implementation is chosen.

Do you propose to add tests for blocks & notifiies, I don't know if this will be easy to test. This probably means lockstep tests with DTLS.

You didn't really give arguments again just adding cleaner observer in matcher just before to add in the `ExchangeStore`. Why does it sound to you a bad idea ?

It's seems to me clearer to put closer the add/clean codes and we are sure that all requests will have its cleanerObserver.

Reply:
> Do you propose to add tests for blocks & notifiies, 

Yes.

> I don't know if this will be easy to test. This probably means lockstep tests with DTLS.

Though the issue is based on the `onError()` or `onSendError()` I would just call them for additional tests. With the secure test you already provided you proofed already your analysis.

Alternatively it may be possible to use the extplugtest BenchmarkClient with the reverse-observe together with DTLS and the NAT to see, if the implementation is robust enough for that scenario.
 
>You didn't really give arguments again just adding cleaner observer in matcher just before to add in the ExchangeStore. Why does it sound to you a bad idea ?

Yes, I didn't provide arguments, because FMPOV your premise with blockwise isn't true. So why change something, if I didn't expect the issue you see? And if you add the cleanup close to the exchange, but the trigger (cancel) is for/from a outer request, then you must forward that trigger. So, this would change some more. 
And so I think, the most important thing is to provide some tests then we see, if it works with the ExchangeCleanupLayer or if a redesign is required, as you propose.

Reply:
> So why change something, if I didn't expect the issue you see?

Because you could find that the code will be better structured and less error prone like this for the future.

>And so I think, the most important thing is to provide some tests then we see, if it works with the ExchangeCleanupLayer or if a redesign is required, as you propose.

I will try to add more tests.